crates/nu-system/src/foreground.rs

```diff
@@ -0,0 +1,53 @@
+// It's a simpler version for fish shell's external process handling.
+//
+// For more information, please check `child_setup_process` function in fish shell.
+// https://github.com/fish-shell/fish-shell/blob/3f90efca38079922b4b21707001d7bb9630107eb/src/postfork.cpp#L140
+#[cfg(target_family = "unix")]
+pub mod external_process_setup {
+    use std::os::unix::prelude::CommandExt;
+    pub fn prepare_to_foreground(external_command: &mut std::process::Command) {
+        unsafe {
+            libc::signal(libc::SIGTTOU, libc::SIG_IGN);
+            libc::signal(libc::SIGTTIN, libc::SIG_IGN);
+
+            external_command.pre_exec(|| {
+                // make the command startup with new process group.
+                // The process group id must be the same as external commands' pid.
+                // Or else we'll failed to set it as foreground process.
+                // For more information, refer to `fork_child_for_process` function:
+                // https://github.com/fish-shell/fish-shell/blob/023042098396aa450d2c4ea1eb9341312de23126/src/exec.cpp#L398
+                libc::setpgid(0, 0);
+                Ok(())
+            });
+        }
+    }
+
+    /// Before call this function, `prepare_to_foreground` function must be called.  Or else
+    /// it'll failed with silence.
+    pub fn set_foreground(process: &std::process::Child) {
```

Comment:
Can we replace this manual API to uphold correctness?
Maybe by making this whole thing a newtype enum around both `std::process::Command` for the unitialized stage and `std::process::Child` for the running version? Could this then be safely handleable with a Drop impl and a way to `.spawn()` the prepared command

Reply:
Oh, thanks for the idea, I like it :)  Added `ForegroundProcess` and `ForegroundChild` for the new code structure